### ISSUE TITLE
Named slots support for popover-menu

### ISSUE THREAD

**Issue by georgehrke** [nextcloud/nextcloud-vue#103]
There are many special needs for the popover-menu (like a color input in calendar or the delete confirmation in tasks) that do not make sense to be implement in vue-components itself.
Therefore popovermenu should allow to hand over named slots to provide individual popover-menu items.

**Comment by skjnldsv** [comment]
I'm actually not really sure about this! :thinking: 
This have crossed my mind a few times, but in the end I don't think this is a good idea to have optional random stuff here.

This goes against the idea of standard. The question we need to really think before is why do you need that in the first place? Why does it look like that? What is the proper UX behind this action.

The example of the task delete confirmation is a good one. Having a countdown AND a confirm makes no sense. Either you have to **click again**, or you have to **wait and cancel**. Usually a second click is a better UX since having to wait is a banned behaviour in the ux guidelines.
Having a menu entry with a first action, then the vue menu mutate to change the text and ask for a confirmation makes perfect sense. A countdown doesn't really have this place here, especially since we do not use it anywhere on nextcloud (though we tend to have pretty destructive actions without rollback, so not really great either :see_no_evil:) 

Regarding the color picker, it might make sense to implement the input support like we do in the standard.
It's fine if we add a colorPicker the same way we have a datePicker in a popover for example :)
I'd say this is a 'pick' action and needs to be elaborated (the input standard already exists [since 14](https://docs.nextcloud.com/server/14/developer_manual/design/popovermenu.html))

**Comment by georgehrke** [comment]
> Having a countdown AND a confirm makes no sense. Either you have to click again

But then Windows users have been trained for years to double click everything they see. So they will immediately click again and never see the warning. Having to wait before you can click again is basically the countdown behaviour we have right now.

> or you have to wait and cancel. Usually a second click is a better UX since having to wait is a banned behaviour in the ux guidelines.

I know that the current behaviour is not optimal, but it's still better than accidentally loosing your entire calendar / address book. The best option would be Undo / Trashbin, but that can only be implemented in the server, not the client. (https://github.com/nextcloud/server/issues/1662, lets talk about this at the hackweek and add it to 16)

> Regarding the color picker, it might make sense to implement the input support like we do in the standard.
It's fine if we add a colorPicker the same way we have a datePicker in a popover for example 😀

That would be great! :)

**Comment by skjnldsv** [comment]
> But then Windows users have been trained for years to double click everything they see. So they will immediately click again and never see the warning. Having to wait before you can click again is basically the countdown behaviour we have right now.

Proper UX required a small delay(debounce) then :)
Even 50ms should be enough. And I would even say that even win users knows not to doubleclick when the hand cursor is showing :see_no_evil: (I hope :pray:) 

> The best option would be Undo / Trashbin, but that can only be implemented in the server, not the client. (nextcloud/server#1662, lets talk about this at the hackweek and add it to 16)

We support [the undo option on the app navigation](https://docs.nextcloud.com/server/14/developer_manual/design/navigation.html#undo-entry) though. Delaying the request with 7second is perfect for that UX :wink: 
You can even add a countdown here if you want (different from the countdown to action, this would be an informative countdown)

**Comment by georgehrke** [comment]
I created a dedicated ticket about the color-input: #138 

Closing this ticket for now.

**Comment by jancborchardt** [comment]
Yup, for the record, especially on those 2 examples you give

> a color input in calendar or the delete confirmation in tasks

– there should be some sort of standardization. :)

**Comment by skjnldsv** [comment]
@jancborchardt About the delete confirmation I had this idea for the undo thing on the lists:
https://github.com/nextcloud/contacts/issues/637

I feel like it looks very nice.
If we want to think about a standard, we should think about various examples (deleted in the appnavigation is already standardised: https://docs.nextcloud.com/server/stable/developer_manual/design/navigation.html#undo-entry

**Comment by jancborchardt** [comment]
@skjnldsv yeah, was thinking about that too, some web apps and maybe some Android apps do this as far as I remember.

The issue is that it still shows the entry for some seconds. It’s important that we offer a means of error prevention, but not sacrificing the default experience.

What about directly vanishing the element (via compress animation), and showing the undo entry as a sticky element at the _bottom_ of the relevant view. So for Talk conversations, it would show on the bottom of `#app-navigation`, for Mails on the bottom of the `#app-content-list`, for Files on the bottom of the file list, and for unsharing on the bottom of the sidebar.

So basically, any sort of listing component needs a specific undo component, which stickies one of the entries to the end. (Just like it should also have an emptycontent component for example?)

That way we tick all the boxes:
- [x] Default UX is still good
- [x] Undo is discoverable, and actually similar like on Android where it’s also at the bottom always
- [x] The undo function is in a relevant place, and not just at the top via the showNotification we are phasing out anyway

**Comment by skjnldsv** [comment]
@jancborchardt what would multiple pending undo looks like then? :)
That is my main focus, any idea?

Could you do a mockup?

**Comment by jancborchardt** [comment]
If we really want to do it, we can simply do it stacked. It would only ever show one undo, but when you deleted 3 files in a row and then undo once, it shows the undo for the middle file again.

But yea – this only if we really want to. :)

**Comment by skjnldsv** [comment]
No, I agree with you, you're right! One undo per *action* is more sense ux wise than one per *element* :)
